plumbing/object/commitgraph/commitnode_graph_builder.go

```diff
@@ -0,0 +1,231 @@
+				return nil, err
+			}
+			commitNode, err := commitIndex.Get(hash)
+			if err != nil {
+				return nil, err
+			}
+			if err := builder.AddCommitNode(commitNode); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	err := heads.ForEach(builder.AddCommit)
+	if err != nil {
+		return nil, err
+	}
+	builder.index.Sort()
+
+	return builder.index, nil
+}
+
+var errNotReady = errors.New("not ready")
+
+type commitGraphBuilder struct {
+	index       *commitgraph.MemoryIndex
+	toWalk      []*object.Commit // Stack of commits being walked - this may get as large as the longest chain
+	commitIndex CommitNodeIndex
+}
+
+func (c *commitGraphBuilder) AddCommitNode(commitNode CommitNode) error {
+	if c.has(commitNode.ID()) {
+		return nil
+	}
+	if isValid(commitNode) {
+		// we can just copy this into the index
+		c.copyCommitNode(commitNode.ID(), commitNode)
+		return nil
+	}
+
+	// Not valid
+	commit, err := commitNode.Commit()
+	if err != nil {
+		return err
+	}
+	return c.AddCommit(commit)
+}
+
+func (c *commitGraphBuilder) AddCommit(commit *object.Commit) error {
+	if c.has(commit.Hash) {
+		return nil
+	}
+
+	c.push(commit)
+	for {
+		commit := c.peek()
+		if commit == nil {
+			break
+		}
+		if err := c.walk(commit); err == errNotReady {
+			continue
+		} else if err != nil {
+			return err
+		}
+		c.pop()
+	}
+	return nil
+}
+
+func (c *commitGraphBuilder) peek() *object.Commit {
+	if len(c.toWalk) == 0 {
+		return nil
+	}
+	return c.toWalk[len(c.toWalk)-1]
+}
+
+func (c *commitGraphBuilder) pop() *object.Commit {
+	if len(c.toWalk) == 0 {
+		return nil
+	}
+	commit := c.toWalk[len(c.toWalk)-1]
+	c.toWalk = c.toWalk[:len(c.toWalk)-1]
+	return commit
+}
+
+func (c *commitGraphBuilder) push(commit *object.Commit) {
+	c.toWalk = append(c.toWalk, commit)
+}
+
+func (c *commitGraphBuilder) has(hash plumbing.Hash) bool {
+	_, err := c.index.GetIndexByHash(hash)
+	return err == nil
+}
+
+func (c *commitGraphBuilder) copyCommitNode(hash plumbing.Hash, commitNode CommitNode) {
+	c.index.Add(hash, &commitgraph.CommitData{
+		TreeHash:     commitNode.TreeHash(),
+		ParentHashes: commitNode.ParentHashes(),
+		Generation:   commitNode.Generation(),
+		GenerationV2: commitNode.GenerationV2(),
+		When:         commitNode.CommitTime().UTC(),
+	})
+}
+
+func isValid(commitNode CommitNode) bool {
+	return commitNode.Generation() != math.MaxUint64 && (commitNode.GenerationV2() != 0 || commitNode.CommitTime().Unix() == 0)
+}
+
+func (c *commitGraphBuilder) walk(commit *object.Commit) error {
+	if c.has(commit.Hash) {
+		// Already in the graph and nothing to do
+		return nil
+	}
+
+	commitNode, err := c.commitIndex.Get(commit.Hash)
+	if err != nil {
+		return err
+	}
+	if isValid(commitNode) {
+		// commitData can be copied in to the graph
+		c.copyCommitNode(commit.Hash, commitNode)
+		return nil
+	}
+
+	// Not in the graph yet
+	// OK so to add a node we need to check its parents
+	// If all of the parents are in the graph then we can add the node.
```

Comment:
```suggestion
	// The node is not in the graph yet. But to add it we first need to check its parents.
	// Only when all of the parents are in the graph that we can add the node.
```

Reply:
Sorry about this I'd done quite a lot of rewriting of the comments - I think things should be a lot clearer in the rewrite!